File: src/walkway_perception/src/obstacle_detector.cpp
#include <algorithm>
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "std_msgs/msg/bool.hpp"
#include "rclcpp/qos.hpp"

class ObstacleDetector : public rclcpp::Node {
    public:
    ObstacleDetector() : Node("obstacle_detector"), danger_threshold_(1.5) {
        subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", rclcpp::SensorDataQoS(), std::bind(&ObstacleDetector::scan_callback, this, std::placeholders::_1)
        );

        publisher_ = this->create_publisher<std_msgs::msg::Bool>("/obstacle_alert", 10);
    }

    private:
    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg) const {
        
        auto start_it = msg->ranges.begin() + 150;
        auto end_it = msg->ranges.begin() + 210;

        auto min_it = std::min_element(start_it, end_it);
        float closest_distance = *min_it;

        auto alert_msg = std_msgs::msg::Bool(); // Create the message object

        if (closest_distance < danger_threshold_) {
            alert_msg.data = true;
            RCLCPP_WARN(this->get_logger(), "OBSTACLE AHEAD! Distance: %.2fm", closest_distance);
        } else {
            alert_msg.data = false;
            RCLCPP_INFO(this->get_logger(), "Path clear.");
        }

        publisher_->publish(alert_msg);
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr publisher_;
    float danger_threshold_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ObstacleDetector>());
    rclcpp::shutdown();
    return 0;
}

File: src/walkway_perception/scripts/obstacle_detector.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool
from rclpy.qos import qos_profile_sensor_data

class ObstacleDetector(Node):
    def __init__(self):
        super().__init__('obstacle_detector')

        # 1. The "Camera" (Subscriber)
        self.subscription_ = self.create_subscription(
            LaserScan, '/scan', self.scan_callback, qos_profile_sensor_data
        )

        # 2. The "Walkie-Talkie" (Publisher)
        self.publisher_ = self.create_publisher(Bool, '/obstacle_alert', 10)

        self.danger_threshold = 1.5 # We consider anything closer than 1.5m a threat

    def scan_callback(self, msg):
        front_cone = msg.ranges[150:210]
        closest_distance = min(front_cone)
        alert_msg = Bool() # Create an empty boolean message

        if closest_distance < self.danger_threshold:
            alert_msg.data = True
            self.get_logger().warn(f"OBSTACLE AHEAD! DISTANCE: {closest_distance:.2f}m")
        else:
            alert_msg.data = False
            self.get_logger().info("Path clear.")
        
        self.publisher_.publish(alert_msg)

def main(args=None):
    rclpy.init(args=args)
    node = ObstacleDetector()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
